core/appmodule/event.go

```diff
@@ -0,0 +1,37 @@
+package appmodule
+
+import (
+	"context"
+
+	"google.golang.org/protobuf/runtime/protoiface"
+)
+
+// HasEventListeners is the extension interface that modules should implement to register
+// event listeners.
+type HasEventListeners interface {
+	AppModule
+
+	// RegisterEventListeners registers the module's events listeners.
+	RegisterEventListeners(registrar *EventListenerRegistrar)
+}
+
+// EventListenerRegistrar allows registering event listeners.
+type EventListenerRegistrar struct {
+	listeners []any
+}
+
+// GetListeners gets the event listeners that have been registered
+func (e *EventListenerRegistrar) GetListeners() []any {
+	return e.listeners
+}
+
+// RegisterEventListener registers an event listener for event type E.
+func RegisterEventListener[E protoiface.MessageV1](registrar *EventListenerRegistrar, listener func(context.Context, E)) {
```

Comment:
Why is this not defined on `EventListenerRegistrar` itself?

Reply:
My understanding is that go doesn't support generic methods

Reply:
I think you can if you define `E protoiface.MessageV1` externally.

E.g.
```go
type E protoiface.MessageV1
```

Reply:
But then it wouldn't be generic